src/main/scala/is/hail/linalg/GridPartitioner.scala

```diff
@@ -133,42 +138,66 @@ case class GridPartitioner(blockSize: Int, nRows: Long, nCols: Long, maybeSparse
-  //   all entries with -lowerBandwidth <= (colIndex - rowIndex) <= upperBandwidth
-  def bandedBlocks(lowerBandwidth: Long, upperBandwidth: Long): Array[Int] = {
-    require(lowerBandwidth >= 0 && upperBandwidth >= 0)
+  //   all elements with lower <= jj - ii <= upper
+  def bandBlocks(lower: Long, upper: Long): Array[Int] = {
+    require(lower <= upper)
     
-    val lowerBlockBandwidth = indexBlockIndex(lowerBandwidth + blockSize - 1)
-    val upperBlockBandwidth = indexBlockIndex(upperBandwidth + blockSize - 1)
+    val lowerBlock = java.lang.Math.floorDiv(lower, blockSize).toInt
+    val upperBlock = java.lang.Math.floorDiv(upper + blockSize - 1, blockSize).toInt
 
     (for { j <- 0 until nBlockCols
-           i <- ((j - upperBlockBandwidth) max 0) to
-                ((j + lowerBlockBandwidth) min (nBlockRows - 1))
+           i <- ((j - upperBlock) max 0) to
+                ((j - lowerBlock) min (nBlockRows - 1))
     } yield (j * nBlockRows) + i).toArray
   }
 
-  // returns increasing array of all blocks intersecting the rectangle [firstRow, lastRow] x [firstCol, lastCol]
-  def rectangularBlocks(firstRow: Long, lastRow: Long, firstCol: Long, lastCol: Long): Array[Int] = {
-    require(firstRow >= 0 && lastRow < nRows)
-    require(firstCol >= 0 && lastCol < nCols)
-    
-    if (firstRow > lastRow || firstCol > lastCol)
-      return Array.empty[Int]
-    
-    val firstBlockRow = indexBlockIndex(firstRow)
-    val lastBlockRow = indexBlockIndex(lastRow)
-    val firstBlockCol = indexBlockIndex(firstCol)
-    val lastBlockCol = indexBlockIndex(lastCol)
+  // returns increasing array of all blocks intersecting the rectangle
+  // [r(0), r(1)) x [r(2), r(3)), i.e. [startRow, stopRow) x [startCol, stopCol)
+  // rectangle checked in Python
+  def rectangleBlocks(r: Array[Long]): Array[Int] = {
+    val startBlockRow = indexBlockIndex(r(0))
+    val stopBlockRow = java.lang.Math.floorDiv(r(1) - 1, blockSize).toInt + 1
+    val startBlockCol = indexBlockIndex(r(2))
+    val stopBlockCol = java.lang.Math.floorDiv(r(3) - 1, blockSize).toInt + 1
     
-    (for { j <- firstBlockCol to lastBlockCol
-           i <- firstBlockRow to lastBlockRow
+    (for { j <- startBlockCol until stopBlockCol
+           i <- startBlockRow until stopBlockRow
     } yield (j * nBlockRows) + i).toArray
   }
 
   // returns increasing array of all blocks intersecting the union of rectangles
-  def rectangularBlocks(rectangles: Array[Array[Long]]): Array[Int] = {
-    require(rectangles.forall(r => r.length == 4))
-    val rects = rectangles.foldLeft(Set[Int]())((s, r) => s ++ rectangularBlocks(r(0), r(1), r(2), r(3))).toArray    
-    scala.util.Sorting.quickSort(rects)
-    rects
+  // rectangles checked in Python
+  def rectanglesBlocks(rectangles: Array[Array[Long]]): Array[Int] = {
+    val blocks = rectangles.foldLeft(Set[Int]())((s, r) => s ++ rectangleBlocks(r)).toArray    
+    scala.util.Sorting.quickSort(blocks)
```

Comment:
why this sort motif?

Reply:
Because using java's quicksort to directly mutate the Array[Int] input with no boxing seems cleaner/faster than `sorted`...even if the latter eventually uses "TimSort" or "miniTimSort" :)
```
  def sorted[B >: A](implicit ord: Ordering[B]): Repr = {
    val len = this.length
    val b = newBuilder
    if (len == 1) b ++= this
    else if (len > 1) {
      b.sizeHint(len)
      val arr = new Array[AnyRef](len)  // Previously used ArraySeq for more compact but slower code
      var i = 0
      for (x <- this) {
        arr(i) = x.asInstanceOf[AnyRef]
        i += 1
      }
      java.util.Arrays.sort(arr, ord.asInstanceOf[Ordering[Object]])
      i = 0
      while (i < arr.length) {
        b += arr(i).asInstanceOf[A]
        i += 1
      }
    }
    b.result()
  }
```

Reply:
I'm not super worried about the extra copy. I'm glad we looked closer ,though, because the line above it is quadratic in the number of rectangles!

Reply:
I don't see why it's quadratic.

I had more verbose optimized versions, but was convinced to back off them:
https://github.com/hail-is/hail/pull/3094

Reply:
set ++ set is linear in the total size of both sets

Reply:
Nm I see why. You’d rather I revert to earlier approach?

Reply:
rectangles.flatMap.distinct.sorted seems like super clear code, and the performance will be fine

Reply:
Wait, isn’t set ++ set just adding elements of second set to first set, with each addition O(1), so independent of size of first set (presuming good hashing). I used sets instead of flatmap for lower memory overhead.

Reply:
`mutable.Set ++= set2` is O(set2.size). Set ++ Set is O(total size), since it needs to be functional (adding can't invalidate the old set, so you need to copy everything)

Reply:
(Not at computer, will look at implementation later)

Reply:
then I’ll switch to mutable set

Reply:
what's wrong with `flatMap.distinct.sorted`?

Reply:
Memory. Won’t matter in practice with default block size but Id like to be robust to filtering using a list of large overlapping rectangles and a smaller block size.